src/utils/images.ts

```diff
@@ -18,3 +24,52 @@ export const getImageDetails = (
     imageDirectory,
   }
 }
+
+/**
+ * Util method to adjust image src in provided HTML, so that they can load from
+ * the correct source (i.e. GitHub or as base64).
+ */
+export const adjustImageSrcs = (
+  html: string,
+  mediaData: MediaData[]
+): string => {
+  const $ = cheerio.load(html)
+
+  $("img").each((_, element) => {
+    const src = $(element).attr("src")
+
+    if (src && isLinkInternal(src)) {
+      $(element).attr(
+        "src",
+        mediaData.find((media) => src === `/${media.mediaPath}`)?.mediaUrl ||
+          "/placeholder_no_image.png"
+      )
+    }
+
+    // Set default placeholder image if image fails to load
+    $(element).attr(
+      "onerror",
+      "this.onerror=null; this.src='/placeholder_no_image.png';"
+    )
+  })
+
+  return $.html()
+}
+
+/**
+ * Util method to extract all image srcs that are internal from the provided HTML.
+ */
+export const getMediaSrcsFromHtml = (html: string): string[] => {
+  const $ = cheerio.load(html)
+  const mediaSrcs: string[] = []
+
+  $("img").each((_, element) => {
+    const src = $(element).attr("src")
+
+    if (src && isLinkInternal(src)) {
```

Comment:
why do we want to introuduce this logic in the first place ah? 
could we not treat this as long as it is an image, we cache it unless it needs to be invalidted?

Reply:
Hmm caching still happens, just on the browser:
1. If it is an external link, the browser will fetch it once and cache it on subsequent re-renders on the preview.
2. If it is a non-private site, the frontend will query the backend, which will return the raw.githubusercontent.com image path, so the browser will fetch it once and cache it on subsequent re-renders on the preview.
3. If it is a private site, the frontend will query the backend, which will return the base64 encoding of the image loaded from GitHub/GGS. The caching is done by react-query using the image src URL specified by the user.

Overall caching will still happen, just that we are tackling the third situation as the browser is spending time converting the raw.githubusercontent.com result into base64.

Reply:
Hmm why depend on the browser here ah? we do have control over the cache header for our base 64, if we are relying on the browser should we then instead just make our cache header longer? 
i think using one solution might be better here than using 2 different solutions for the same problem!

Reply:
The two solutions is because one is a URL and the other is a base64 image. If the image src is a URL, the browser automatically does this caching for us, and we should leave it to the browser to determine their caching strategy, etc. depending on the user's preferences too.

In the other case, we are loading a base64 image, which by definition is not cached (since it is not loading the image from any external source, so there is nothing to cache here). The problem originally is because we always regenerate the base64 on every render, which because there is no caching involved here, the preview rendering is slow. Instead, we rely on react-query to perform the caching on the relative image URL as the key and offload the base64 generation to the backend, so we don't need to keep regenerating the base64 on every render unless the user made a change to the image URL, which is the key used for the cache.

Essentially they are two separate problems, so two separate solutions are needed.